tests/test_integrals.py

```diff
@@ -688,3 +718,648 @@ def test_biest_interpolators(self):
+        intersect = intersect[intersect >= k[2]]
+        np.testing.assert_allclose(z1, intersect[[0, 2, 4]], rtol=1e-6)
+        np.testing.assert_allclose(z2, intersect[[0, 3, 5]], rtol=1e-6)
+
+    @pytest.mark.unit
+    def test_extrema_first_and_before_z2(self):
+        """Test that bounce points are computed correctly."""
+        start = -1.2 * np.pi
+        end = -2 * start + 1
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 10,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 10,
+        )
+        dB_dz = B.derivative()
+        pitch = 1 / B(dB_dz.roots(extrapolate=False))[1] + 1e-13
+        z1, z2 = bounce_points(pitch, k, B.c, dB_dz.c, check=True)
+        z1, z2 = TestBounce1DPoints.filter(z1, z2)
+        assert z1.size and z2.size
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(z1[0], -0.671904, rtol=1e-6)
+        np.testing.assert_allclose(z1, intersect[[0, 3, 5]], rtol=1e-5)
+        # intersect array could not resolve double root as single at index 0,1
+        np.testing.assert_allclose(intersect[0], intersect[1], rtol=1e-5)
+        np.testing.assert_allclose(z2, intersect[[2, 4, 6]], rtol=1e-5)
+
+    @pytest.mark.unit
+    def test_get_extrema(self):
+        """Test computation of extrema of |B|."""
+        start = -np.pi
+        end = -2 * start
+        k = np.linspace(start, end, 5)
+        B = CubicHermiteSpline(
+            k, np.cos(k) + 2 * np.sin(-2 * k), -np.sin(k) - 4 * np.cos(-2 * k)
+        )
+        dB_dz = B.derivative()
+        ext, B_ext = _get_extrema(k, B.c, dB_dz.c)
+        mask = ~np.isnan(ext)
+        ext, B_ext = ext[mask], B_ext[mask]
+        idx = np.argsort(ext)
+
+        ext_scipy = np.sort(dB_dz.roots(extrapolate=False))
+        B_ext_scipy = B(ext_scipy)
+        assert ext.size == ext_scipy.size
+        np.testing.assert_allclose(ext[idx], ext_scipy)
+        np.testing.assert_allclose(B_ext[idx], B_ext_scipy)
+
+
+def _mod_cheb_gauss(deg):
+    x, w = chebgauss(deg)
+    w /= chebweight(x)
+    return x, w
+
+
+def _mod_chebu_gauss(deg):
+    x, w = roots_chebyu(deg)
+    w *= chebweight(x)
+    return x, w
+
+
+class TestBounce1DQuadrature:
+    """Test bounce quadrature."""
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize(
+        "is_strong, quad, automorphism",
+        [
+            (True, tanh_sinh(40), None),
+            (True, leggauss(25), "default"),
+            (False, tanh_sinh(20), None),
+            (False, leggauss_lob(10), "default"),
+            # sin automorphism still helps out chebyshev quadrature
+            (True, _mod_cheb_gauss(30), "default"),
+            (False, _mod_chebu_gauss(10), "default"),
+        ],
+    )
+    def test_bounce_quadrature(self, is_strong, quad, automorphism):
```

Comment:
@f0uriest
> Main question I forgot to include in the actual review comment:
> 
> There are a few different automorphisms (sin, arcsin) and different quadrature methods (gauss-legendre, gauss-chebyshev, tanh-sinh). Is there a clear winning combination? Or is it still problem dependent?

Yes, the clear winner is the default choice, which is Gauss-Legendre after applying the `sin` automorphism. The `arcsin` one adds a `1/(1-x^2)` singularity that makes things bad for Gauss quadrature. Chebyshev quadrature is not suited for these integrals. Either way, removing the singularity via change of variables always outperforms Tanh-Sinh quadrature in my empirical testing converging to the true answer at lower resolution.

In theory `arcsin` should have made things better for `tanh_sinh` quadrature as it trades more singular (which the tanh_sinh performs better at) for less floating point error (since it pulls quadrature nodes away from the singularity where floating point errors are worst), but my testing showed that `tanh_sinh` simply does not converge well, regardless of whether a change of variables is used or not.

I thought `tanh_sinh` must be hitting a floating point error plateau, but back during the semester I also used `tanh_sinh` with arbitrary precision floating point arithmetic using mpl, and it sill could not converge past a certain point. At low number of quadrature points ~10, it can sometimes get to about 5e-5 error, but as you increase the number of quadrature points to ~20, 50, or 400 the error kicks up to 1e-3. (The Gauss ones converges exponentially after the change of variables).

Some tests used for these results are
- ``test_bounce_quadrature``
- ``elliptic_incomplete``
- ``test_binormal_drift_bounce1d``

The test ``elliptic_incomplete`` is nice because it is independent of all the bounce integrals and splining. Onec can just change the `fixed_elliptic` function above to using tanh_sinh arcsin or regular tanh-sinh to see it perform worse.